oak_runtime/introspection_browser_client/components/Root/index.tsx

```diff
@@ -0,0 +1,216 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import React from 'react';
+import ApplicationStateOverview from '~/components/ApplicationStateOverview';
+import EventList from '~/components/EventList';
+import introspectionEventsProto from '~/proto/introspection_events_pb';
+
+// Requests the list of introspection events provided by the Oak runtime's
+// auxiliary introspection server.
+function loadSerializedEvents(): Promise<Uint8Array> {
+  return new Promise((resolve, reject) => {
+    const eventsRequest = new XMLHttpRequest();
+    eventsRequest.open(
+      'GET',
+      'http://localhost:1909/introspection-events',
+      true
+    );
+    eventsRequest.responseType = 'arraybuffer';
+
+    eventsRequest.addEventListener('load', () =>
+      resolve(eventsRequest.response)
+    );
+    eventsRequest.addEventListener('error', reject);
+    eventsRequest.addEventListener('abort', reject);
+
+    eventsRequest.send(null);
+  });
+}
+
+export interface OakApplicationState {
+  nodeInfos: NodeInfos;
+  channels: Channels;
+}
+
+type NodeId = number;
+type AbiHandle = number;
+type NodeInfos = Map<NodeId, NodeInfo>;
+interface NodeInfo {
+  abiHandles: Map<AbiHandle, ChannelHalf>;
+}
+
+type ChannelID = number;
+enum ChannelHalfDirection {
+  Read,
+  Write
+}
+interface ChannelHalf {
+  channelId: ChannelID;
+  direction: ChannelHalfDirection;
+}
+interface Message {
+  data: Uint8Array;
+  channels: ChannelHalf[];
+}
+interface Channel {
+  id: ChannelID;
+  messages: Message[];
+}
+type Channels = Map<ChannelID, Channel>;
+
+function eventReducer(
+  applicationState: OakApplicationState,
+  event: introspectionEventsProto.Event
+) {
+  const eventType = event.getEventDetailsCase();
+  const { EventDetailsCase } = introspectionEventsProto.Event;
+
+  switch (eventType) {
+    case EventDetailsCase.NODE_CREATED:
+      applicationState.nodeInfos.set(event!.getNodeCreated()!.getNodeId(), {
+        abiHandles: new Map()
+      });
+
+      break;
```

Comment:
Are these breaks actually necessary? Does typescript have implicit fall-through?

Reply:
Sadly yes

Reply:
![image](https://user-images.githubusercontent.com/8875406/92456950-7e65ca80-f1bb-11ea-8d77-77f26596e380.png)
Cursed JavaScript behaviour

Reply:
The stuff of nightmares

Reply:
![image](https://user-images.githubusercontent.com/8875406/92459024-119fff80-f1be-11ea-9fcc-b9e78b94fec0.png)